### PR TITLE
Iterating over multiple pairs of PET and T1w files

### LINKED ISSUE #5: [BUG][Feature Request] Handle Multiple T1w's
We currently run into errors when processing multiple T1w's

See attached log file:

[expected_path_got_list.log](https://github.com/bendhouseart/petdeface/files/11626001/expected_path_got_list.log)

### LINKED ISSUE #6: [BUG] Nipype Error on Non-existent T1w
note that nipype is encountering errors when looking for t1w files that don't exist,
e.g.  in the naming scheme for ds004230 the anat folders (containing t1w's) are only within
the baseline session. Errors are raised for T1w's not existing for the rescan session, see below:

```bash
230602-09:29:36,844 nipype.workflow ERROR:
         Node deface_t1w.a01 failed to run on host MH02276145MLI.
230602-09:29:36,844 nipype.workflow ERROR:
         Saving crash info to /Users/galassiae/Projects/petdeface/crash-20230602-092936-galassiae-deface_t1w.a01-259d03d6-4e35-4b40-b7bd-0cfdbf270296.pklz
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 816, in move
    os.rename(src, real_dst)
FileNotFoundError: [Errno 2] No such file or directory: '/Users/galassiae/Data/pet_defacing_data/ds004230_faced/deface_pet_workflow/_session_id_rescanbrain_subject_id_PS19/deface_t1w/_0x5e1eb1e0e25d19c963fbd7a0618ce8d7_unfinished.json' -> '/Users/galassiae/Data/pet_defacing_data/ds004230_faced/deface_pet_workflow/_session_id_rescanbrain_subject_id_PS19/deface_t1w/_0x5e1eb1e0e25d19c963fbd7a0618ce8d7.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/galassiae/Library/Caches/pypoetry/virtualenvs/petdeface-glZEKbHu-py3.10/lib/python3.10/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/Users/galassiae/Library/Caches/pypoetry/virtualenvs/petdeface-glZEKbHu-py3.10/lib/python3.10/site-packages/nipype/pipeline/engine/nodes.py", line 542, in run
    shutil.move(hashfile_unfinished, hashfile_unfinished.replace("_unfinished", ""))
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 836, in move
    copy_function(src, real_dst)
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 434, in copy2
    copyfile(src, dst, follow_symlinks=follow_symlinks)
  File "/opt/homebrew/Cellar/python@3.10/3.10.11/Frameworks/Python.framework/Versions/3.10/lib/python3.10/shutil.py", line 254, in copyfile
    with open(src, 'rb') as fsrc:
FileNotFoundError: [Errno 2] No such file or directory: '/Users/galassiae/Data/pet_defacing_data/ds004230_faced/deface_pet_workflow/_session_id_rescanbrain_subject_id_PS19/deface_t1w/_0x5e1eb1e0e25d19c963fbd7a0618ce8d7_unfinished.json'
```

Hopefully this is as simple as a fix as popping sessions/folders without t1w's from the list of sessions to be processed and/or filtering the bids/pybids query a bit better.

### LINKED ISSUE #8: [ENH]: consider to refactor data grabbing
We should refactor the data grabbing to be more in line with fmriprep, but also as discussed here: https://neurostars.org/t/nipype-how-can-i-iterate-over-subjects-and-then-dynamically-over-sessions/18451

### PR DESCRIPTION
**Pull request by bilgelm**
Fixes #6 by throwing `RuntimeError` if there are no T1w images for a given participant.

Resolves #8 by implementing `init_single_subject_wf`. (This also fixes #5.) I couldn't get this approach to work with `create_weighted_average_pet` and `create_apply_str`, so I wrote the wrappers `WeightedAverage` and `ApplyMideface`. I replaced `shutil` file copying with nipype's `DataSink`, but only the "derivatives" option of `PetDeface`'s `placement` argument is currently implemented.

Minor bug fix for weighted average PET: previous implementation used the sum of mid time points in the denominator, but this should instead be the range of mid time points (i.e., the duration over which TAC is specified). This change should not affect PET-MR coregistration as mutual information is invariant to scaling.

This pull request simplifies the dependency list by omitting redundant modules from `pyproject.toml`.

This pull request also implements code formatting with black, flake8, and isort, and sets up pre-commit hooks for these formatting checks.

### CODE REVIEW

**Review by mnoergaard — COMMENTED** [review]
Many thanks for making this PR, @bilgelm! I am still not quite convinced that the updated data grabbing is the most ideal (I would rather try to make a main workflow grabbing subjects, and then a subsequent subject workflow grabbing all sessions. Because there you can grab just a single MR for a given subject, and then use this to iterate over multiple PET sessions. It also accommodates the inheritance principle). Nevertheless, I am quite happy with the remaining parts of your PR, so many thanks for these edits/comments!

File: petdeface/pet.py
Comment (bendhouseart) [inline petdeface/pet.py:33]: What happened here? @mnoergaard

File: petdeface/petdeface.py
Comment (bendhouseart) [inline petdeface/petdeface.py:20]: I know this stuff is goofy, but removing it breaks this when running it in development mode, see:

```(petdeface-py3.10) petdeface % poetry install --sync
Configuration file exists at /Library/Preferences/pypoetry, reusing this directory.

Consider moving TOML configuration files to /Library/Application Support/pypoetry, as support for the legacy directory will be removed in an upcoming release.
Installing dependencies from lock file

Package operations: 17 installs, 4 updates, 1 removal

  • Removing bids (0.0)
  • Updating numpy (1.24.4 -> 1.25.2)
  • Updating setuptools (67.8.0 -> 68.1.2)
  • Installing distlib (0.3.7)
  • Installing lazy-loader (0.3)
  • Installing platformdirs (3.10.0)
  • Updating tifffile (2023.7.10 -> 2023.8.25)
  • Installing cfgv (3.4.0)
  • Installing identify (2.5.27)
  • Installing mccabe (0.7.0)
  • Installing mypy-extensions (1.0.0)
  • Installing nodeenv (1.8.0)
  • Installing pathspec (0.11.2)
  • Installing pycodestyle (2.11.0)
  • Installing pyflakes (3.1.0)
  • Updating scikit-image (0.19.3 -> 0.21.0)
  • Installing tomli (2.0.1)
  • Installing virtualenv (20.24.3)
  • Installing black (23.7.0)
  • Installing flake8 (6.1.0)
  • Installing isort (5.12.0)
  • Installing pre-commit (3.3.3)

Installing the current project: petdeface (0.0.0)
(petdeface-py3.10) petdeface % python petdeface/petdeface.py 
Traceback (most recent call last):
  File "/Projects/petdeface/petdeface/petdeface.py", line 20, in <module>
    from .mideface import Mideface
ImportError: attempted relative import with no known parent package```

Comment (bilgelm) [reply petdeface/petdeface.py:20]: Hi @bendhouseart, running `python petdeface/petdeface.py` does give me the same error. Instead of this, would using the petdeface cli within the virtualenv set up by poetry be an acceptable alternative? i.e., `poetry run petdeface -h`. Since petdeface module is installed within this virtualenv, there is no `ImportError`. Otherwise, I can revert this edit.

File: petdeface/petdeface.py
Comment (bilgelm) [reply petdeface/petdeface.py:167]: Thanks! I'll update the niworkflows requirement in `pyproject.toml` so that it is installed from GitHub to capture this change.

File: petdeface/petdeface.py
Comment (bendhouseart) [reply petdeface/petdeface.py:20]: It's solely for debugging and while a try catch doesn't seem all that elegant I've found that it works pretty well. Perhaps I shouldn't misinterpret a failure of my footgun to fire as it being safe, but I've never really bothered to find a better way to handle imports that make packaging and using an ide's debugger work harmoniously.

Let's kick the can and do proper imports in a smaller atomic PR after merging this one.

File: pyproject.toml
Comment (bendhouseart) [inline pyproject.toml:25]: Also, what happened to this include and the rest of the requirements?

Comment (bilgelm) [reply pyproject.toml:25]: They are all requirements of niworkflows, so they are implicitly included.

Comment (bendhouseart) [reply pyproject.toml:25]: Include is required for packaging and pushing to pypi, see [here](https://python-poetry.org/docs/pyproject/#include-and-exclude). And is setup tools included too? So long as it builds and packages itself when your run `poetry build` and `poetry publish`.

File: petdeface/petdeface.py
Comment (bilgelm) [reply petdeface/petdeface.py]: I think my latest commit achieves the desired behavior. My testing has been limited, so hopefully it covers all possible situations 🤞.

There are ambiguous cases. Let's say a subject has t1w for sessions 1, 2, and 4. Which t1w should be used for pet in session 3? I didn't add anything in my code to "break ties" and favor one of these sessions over the others. (In current implementation, all of these t1ws are tied for best match, and the code would pick ses-01_t1w to process ses-03_pet just because it comes first in the list.)

File: petdeface/pet.py
Comment (bilgelm) [reply petdeface/pet.py]: According to [numpy.trapz](https://numpy.org/doc/stable/reference/generated/numpy.trapz.html) documentation, `dx` should be a scalar, and I'm not sure what was happening with `dx` being set to an array. Instead, we can specify the location of the data points using the `x` argument, which accepts arrays. The indexing on `data` was redundant, and so was the `axis=3` argument to `np.trapz`, since the default is `axis=-1`. Denominator should be the range of the TAC, not the sum of mid frame time points.

Comment (mnoergaard) [reply petdeface/pet.py]: Please note that this is a time weighted average (weighted by the frame length) i.e. $W\ =\ \frac{\sum_{i=1}^nw_i\cdot X_i}{\sum_{i=1}^nw_i}$, which matches the formula.

File: petdeface/petdeface.py
Comment (mnoergaard) [reply petdeface/petdeface.py]: Awesome - thanks @bilgelm! @bendhouseart we should go ahead and make the test cases to verify that this indeed covers the different cases. Regarding the ambiguous cases, workflows like fmriprep and smriprep go ahead an create a single anatomical image using mri_robust_register if more than one anatomical file is available. Difficult to say what would be the optimal usage here, but I would be fine if there is match within a session between PET and anatomical then use that, and if there is no anatomical present, then use the first available anatomical.

File: petdeface/pet.py
Comment (bendhouseart) [reply petdeface/pet.py]: You know what, I'm just going to let you two determine which is correct as `np.sum(midframes)` looks very different to me from`midframes[-1] - midframes[0]`. I could walk through this with the debugger, but let's be realistic here.

**Comment by bendhouseart** [comment]
I'm happy with this barring the committed changes to the `pyproject.toml`, and as far as the formula I commented on above I defer to you two when it comes to anything relating to TAC's.

File: petdeface/pet.py
Comment (bendhouseart) [reply petdeface/pet.py]: Okay, ah, ah, laying out $w_i$ made this much more clear, thank you.

**Review by bendhouseart — APPROVED** [review]
LGTM

File: petdeface/pet.py
Comment (mnoergaard) [reply petdeface/pet.py]: I stand corrected - thank you, Murat! I had originally re-used from here: https://github.com/openneuropet/PET_pipelines/blob/main/pyPetSurfer/pet_surfer.py

But going back and looking at my previous matlab implementation I have this:

function create_tac_wavg(tacpath,time,out)

tac=MRIread(tacpath);
tac.vol=trapz(time,tac.vol,4)/(time(end)-time(1));
MRIwrite(tac,out);

**Comment by mnoergaard** [comment]
This also looks good to me! Thank you so much, Murat! We should, however, still move ahead and create the necessary tests and include them in the repo.

**Review by mnoergaard — APPROVED** [review]

**Comment by bendhouseart** [comment]
> This also looks good to me! Thank you so much, Murat! We should, however, still move ahead and create the necessary tests and include them in the repo.

Do you want to merge this or create the tests first? We're going to be limited to really simple unit tests unless do one or more of the following: 

1) set up a self hosted runner  
2) collect faced data that we can distribute off premises  
3) downsample an open faced (hehe) MR image until it looks as good as a pet image or something.

**Pull request merged**
